### PR TITLE
Hierarchies with multiple parents

### PR DESCRIPTION
**Pull request by HeikoTheissen**
- [x] Multiple parents per node are allowed, leading to duplicate nodes in the result set.
- [x] In the result set, `Aggregation.UpPath` annotation contains the node identifiers of one sequence of parents (or ancestors).
- [x] The optional parameter $S$ is removed from `ancestors` and `descendants` (see oasis-tcs/odata-abnf#100).
- [x] The optional parameter $S$ in `traverse` defines the start nodes for the traversal (default is start node = root node).
- [x] The optional parameter $S$ in `rolluprecursive` is needed for rolling up "actual totals" (as opposed to "visual totals", see example 68).

### CODE REVIEW

File: odata-data-aggregation-ext/7 Examples.md [hunk 1]
```diff
@@ -1386,24 +1386,70 @@ Content-Type: application/json
...
+EMEA Central|EMEA
+Phobos|Mars
+Venus|
+
+Then the entity Mars has no node identifier and hence is no node in the hierarchy, even though it occurs as a parent. Its child Phobos cannot be reached from the root Sales and hence is an orphan, as is Venus.
+
+Mars and Phobos can be made descendants of the root node by adding a relationship:
 ```json
-PATCH /service/SalesOrganizations('Mars')
+PATCH /service/SalesOrganizations('Mars')/Relations
```
Comment (ralfhandl) [inline odata-data-aggregation-ext/7 Examples.md:1452]:
```suggestion
POST /service/SalesOrganizations('Mars')/Relations
```

File: odata-data-aggregation-ext/7 Examples.md [hunk 2]
```diff
@@ -1345,6 +1345,117 @@ ancestors(
...
+    <ReferentialConstraint Property="NodeID"
+                           ReferencedProperty="NodeID" />
+  </NavigationProperty>
+  <Annotation Term="Aggregation.RecursiveHierarchy"
+              Qualifier="MultiParentHierarchy">
+    <Record>
+      <PropertyValue Property="NodeProperty"
+                     PropertyPath="NodeID" />
+      <PropertyValue Property="ParentNavigationProperty"
+                     PropertyPath="Relations/Superordinate" />
```
Comment (HeikoTheissen) [inline odata-data-aggregation-ext/7 Examples.md:1426]: This navigation property path consists of a collection-valued navigation property followed by a single-valued one.

File: odata-data-aggregation-ext/5 Vocabulary for Data Aggregation.md [hunk 3]
```diff
@@ -174,39 +174,42 @@ The term `LeveledHierarchy` MUST be applied with a qualifier that can be used to
...
+- additional parameters, depending on the type of test (see below)
 - a Boolean return value for the outcome of the test.
 
 The following functions are defined:
-- `isroot` tests if the given entity is a root of the hierarchy
-- `isdescendant` tests if the given entity is a descendant of an ancestor node (whose node identifier is given in a parameter `Ancestor`) with a maximum distance `MaxDistance`, or equals the ancestor if `IncludeSelf` is true
-- `isancestor` tests if the given entity is an ancestor of a descendant node (whose node identifier is given in a parameter `Descendant`) with a maximum distance `MaxDistance`, or equals the descendant if `IncludeSelf` is true
-- `issibling` tests if the given entity and another entity (whose node identifier is given in a parameter `Other`) have the same parent node or both are roots, but are not the same
-- `isleaf` tests if the given entity is without descendants.
+- `isroot` tests if the given entity is a root node of the hierarchy.
```
Comment (HeikoTheissen) [inline odata-data-aggregation-ext/5 Vocabulary for Data Aggregation.md:201]: Do we want to test for root nodes or for start nodes?

File: odata-data-aggregation-ext/6 Hierarchical Transformations.md [hunk 4]
```diff
@@ -273,26 +276,103 @@ results in
...
+Given a start node $x$, let $ρ_0(x)$ be the node $x$ with the annotation $ρ_0(x)/\hbox{\tt @Aggregation.UpPath}\#Q$ set to an empty collection.
 
-More precisely, a _path-to-the-root_ is a node $x$ that is annotated with the term `UpNode` from the `Aggregation` vocabulary [OData-VocAggr](#ODataVocAggr) where the annotation value is the parent node $y$ such that $R(x)$ appears on the right-hand side of the recursive formula for $R(y)$. The annotation value $y$ is again annotated with `Aggregation.UpNode` and so on until a root is reached. Every instance in the output set of `traverse` is related to one path-to-the-root.
+If the `Aggregation.UpPath` annotation of $y$ contains the node identifier of $y$, a cycle has been detected and $ρ(y,x)$ is additionally annotated with term `Aggregation.Cycle`, qualifier $Q$ and value true. The algorithm does then not process the children of this node again.
```
[the remark this answers is not in the document]
Comment (HeikoTheissen) [reply odata-data-aggregation-ext/6 Hierarchical Transformations.md:319]: Assume A > B > C > D and D > C where ">" means "parent of". Then the entry
```json
{"NodeID": "C", "@Aggregation.Cycle": true, "@Aggregation.UpPath": ["D", "C", "B", "A"]}
```
makes clear there is a cycle C-D-C. Without that entry, the cycle would be unclear.

**Review by ralfhandl — APPROVED** [review]

File: odata-data-aggregation-ext/7 Examples.md [hunk 5]
```diff
@@ -1345,6 +1345,117 @@ ancestors(
...
+Venus|
+
+Then the entity Mars has no node identifier and hence is no node in the hierarchy, even though it occurs as a parent. Its child Phobos cannot be reached from the root Sales and hence is an orphan, as is Venus.
+
+Mars and Phobos can be made descendants of the root node by adding a relationship. Note the first segment of the `ParentNavigationProperty` appears at the end of the resource path and the second segment appears in the payload before the `@bind`:
+```json
+POST /service/SalesOrganizations('Mars')/Relations
+Content-Type: application/json
+
+{ "Superordinate@bind": "SalesOrganizations('Sales')" }
```
Comment (HeikoTheissen) [inline odata-data-aggregation-ext/7 Examples.md:1455]: Is it possible to achieve the same effect with a request where both segments appear together?

File: odata-data-aggregation-ext/5 Vocabulary for Data Aggregation.md [hunk 6]
```diff
@@ -174,39 +174,54 @@ The term `LeveledHierarchy` MUST be applied with a qualifier that can be used to
...
+The recursive hierarchy is described in the model by an annotation of the entity type with the complex term `RecursiveHierarchy` with these properties:
+- The `NodeProperty` allows identifying a node in the hierarchy. It MUST be a path with single-valued segments ending in a primitive property. This property holds the node identifier of the node in the hierarchy. Entities for which this path evaluates to null are not nodes of the hierarchy.
+- The `ParentNavigationProperty` allows navigation to the instance or instances representing the parent nodes. It MUST be a collection-valued or nullable single-valued navigation property path that addresses the entity type annotated with this term.
```
Comment (HeikoTheissen) [inline odata-data-aggregation-ext/5 Vocabulary for Data Aggregation.md:187]: Give an example for cycles without temporality.

**Review by GeraldKrause — CHANGES_REQUESTED** [review]

File: odata-data-aggregation-ext/6 Hierarchical Transformations.md [hunk 7]
```diff
@@ -273,32 +280,115 @@ results in
...
+Given a node $x$ annotated with $x/@\hbox{\tt Aggregation.UpPath}\#Q=[x_1,…,x_d]$, where $d≥0$, and given a child $y$ of $x$, let $ρ(y,x)$ be the node $y$ with the annotation
+$$ρ(y,x)/@\hbox{\tt Aggregation.UpPath}\#Q=[{\tt cast}(x[q],\hbox{\tt Edm.String}),x_1,…,x_d].$$
 
-If $h={\tt postorder}$, then
+If the string value of the node identifier of $y$ is among the values on the right-hand side of the previous equation, a cycle has been detected and $ρ(y,x)$ is additionally annotated with
+$$ρ(y,x)/@\hbox{\tt Aggregation.Cycle}\#Q={\tt true}.$$
+The algorithm does then not process the children of this node again.
+
+::: example
+⚠ Example ##ex: If the child of Atlantis is also its parent:
```
[the remark this answers is not in the document]
Comment (HeikoTheissen) [reply odata-data-aggregation-ext/6 Hierarchical Transformations.md:330]: What's wrong with the example?

File: odata-data-aggregation-ext/5 Vocabulary for Data Aggregation.md [hunk 8]
```diff
@@ -174,39 +174,47 @@ The term `LeveledHierarchy` MUST be applied with a qualifier that can be used to
 
 ### ##subsubsec Recursive Hierarchy
 
-A recursive hierarchy organizes entities of a collection as nodes of one or more tree structures. This structure does not need to be as uniform as a leveled hierarchy. It is described by a complex term `RecursiveHierarchy` with these properties:
-- The `NodeProperty` allows identifying a node in the hierarchy. It MUST be a path with single-valued segments ending in a primitive property.
-- The `ParentNavigationProperty` allows navigation to the instance or instances representing the parent nodes. It MUST be a collection-valued or nullable single-valued navigation property path that addresses the entity type annotated with this term. Nodes MUST NOT form cycles when following parent navigation properties.
+A recursive hierarchy is defined on a collection of entities by
+- associating with every entity zero or more entities from the same collection, called its _parents_, and
+- identifying entities that are part of the recursive hierarchy through a single primitive property called the _node identifier_. The entities in the hierarchy are called _nodes_.
```
[the remark this answers is not in the document]
Comment (HeikoTheissen) [reply odata-data-aggregation-ext/5 Vocabulary for Data Aggregation.md]: Check again after my latest commit.

File: odata-data-aggregation-ext/6 Hierarchical Transformations.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (GeraldKrause) [reply odata-data-aggregation-ext/6 Hierarchical Transformations.md:330]: I got confused and thought "its" relates to the "child of Atlantis". 
```suggestion
⚠ Example ##ex: If the child of Atlantis is also the parent of Atlantis:
```

**Review by GeraldKrause — APPROVED** [review]
Approved with the changes discussed in TC call 2023-06-28

**Pull request merged**